### DISCUSSION TITLE
Case Not Being Created - Two Fold, Swap Case For Custom Voicemail Object

### DISCUSSION THREAD

**Discussion by ShinobiSec** [amazon-connect/amazon-connect-salesforce-scv#81] [Q&A] [answered]
There is an error taking place when the packager attempts to create a case. Can anyone assist on this? 

START RequestId: b4998231-106c-4c6b-b9fe-a63d86b8baea Version: $LATEST
[INFO] 2022-01-12T20:41:01.162Z b4998231-106c-4c6b-b9fe-a63d86b8baea Found credentials in environment variables.
[ERROR] 2022-01-12T20:41:10.284Z b4998231-106c-4c6b-b9fe-a63d86b8baea list indices must be integers or slices, not str
[ERROR] 2022-01-12T20:41:10.285Z b4998231-106c-4c6b-b9fe-a63d86b8baea Record 1 Result: Failed to create case
END RequestId: b4998231-106c-4c6b-b9fe-a63d86b8baea
REPORT RequestId: b4998231-106c-4c6b-b9fe-a63d86b8baea Duration: 9378.65 ms Billed Duration: 9379 ms Memory Size: 128 MB Max Memory Used: 96 MB Init Duration: 399.51 ms
START RequestId: ef653d19-79cc-4d79-a17e-6f77095888f8 Version: $LATEST
[ERROR] KeyError: 'Records' Traceback (most recent call last):   File "/var/task/awsscv_vmx_packager.py", line 35, in lambda_handler     for record in event['Records']:
END RequestId: ef653d19-79cc-4d79-a17e-6f77095888f8
REPORT RequestId: ef653d19-79cc-4d79-a17e-6f77095888f8 Duration: 31.03 ms Billed Duration: 32 ms Memory Size: 128 MB Max Memory Used: 96 MB

Once we get everything working, I've created a "Voicemail" object that I hope to use instead of the Case object. We cannot use cases for voicemails in our org. I've added the fields to the Voicemail object exactly the same way they were added to the case object per the Voicemail Express Prerequisites documentation. From what I can gather, we cannot view or edit code any deeper than the lambda function python in the scvvmxpackager to define what object is created. The code I intend to edit is:

  # Create a case in Salesforce
        try:
            if loaded_tags['vm_queue_type'] == 'agent':
                data = {
                    'Subject': vm_prefix + ' voicemail for: ' + entity_name + ', from : ' + international_phone,
                    'Description': 'Voicemail transcript: ' + transcript_contents,
                    'Status': 'New',
                    'Priority': loaded_tags['vm_priority'],
                    'Origin': 'Phone',
                    'OwnerId': sf_agent_id,
                    os.environ['sf_case_vm_attributes']: contact_attributes,
                    os.environ['sf_case_vm_phone_field']: international_phone,
                    os.environ['sf_case_vm_field']: formatted_url
                }

            else:
                data = {
                    'Subject': vm_prefix + ' voicemail for: ' + entity_name + ', from : ' + international_phone,
                    'Description': 'Voicemail transcript: ' + transcript_contents,
                    'Status': 'New',
                    'Priority': loaded_tags['vm_priority'],
                    'Origin': 'Phone',
                    os.environ['sf_case_vm_attributes']: contact_attributes,
                    os.environ['sf_case_vm_phone_field']: international_phone,
                    os.environ['sf_case_vm_field']: formatted_url
                }
            logger.debug(data)

            response = sf.create(sobject='case', data=data)

            logger.info('Record {0} Result: Voicemail created [{1}]'.format(loop_counter, response))
            sf_case = response

        except Exception as e:
            logger.error(e)
            logger.error('Record {0} Result: Failed to create case'.format(loop_counter))
            continue

I'm not sure if the os.environ['sf_case_vm_attributes']: contact_attributes, need adjusted to sf_Voicemail__c_vm_attributes for example, or if it's as simple as changing  response = sf.create(sobject='case', data=data) to response = sf.create(sobject='Voicemail__c', data=data)

Thanks in advance for any help!

**Comment by dougjaso** [comment]
You're mixing up things. os.environ['sf_case_vm_attributes'] is referencing the custom field created on the case object that will be used to store a dump of the contact attributes. Whatever field you want to place the attribute list in, that is the value that should be set. Same goes for os.environ['sf_case_vm_phone_field'] (custom phone field in the case object), and os.environ['sf_case_vm_field'] ( the presigned URL button link). Those fields need to exist in your object. Once you have those in your object, you can use whatever object type you like by switching soobject='case' to be the api name for whatever object you want to use.

**Reply by ShinobiSec** [reply]
Do you have any thoughts on these errors occurring from the deployment followed by the book to the T? I want to make sure it's working before I introduce any code changes. 

[ERROR] 2022-01-12T20:41:10.284Z b4998231-106c-4c6b-b9fe-a63d86b8baea list indices must be integers or slices, not str
[ERROR] 2022-01-12T20:41:10.285Z b4998231-106c-4c6b-b9fe-a63d86b8baea Record 1 Result: Failed to create case 

[ERROR] KeyError: 'Records' Traceback (most recent call last):   File "/var/task/awsscv_vmx_packager.py", line 35, in lambda_handler     for record in event['Records']:


Based on my research this is a code-based error or something in previous execution that is not providing the proper input to the packager. Thus, causing a traceback error. Though, I still cannot see any way to resolve the list indices type error.

**Reply by ShinobiSec** [reply]
Doug, 

You stated that "Whatever field you want to place the attribute list in, that is the value that should be set. Same goes for os.environ['sf_case_vm_phone_field'] (custom phone field in the case object)" 

If I'm understanding correctly, the Voicemail Express Prerequisites doc instructs us to set field labels "Voicemail Link", "Callback Phone", and "Connect Contact Attributes". 

In the case of using the custom object API name Voicemail__c along with the field created per the instructions, wouldn't we want it to be written as such:
 
os.environ['sf_Voicemail__c_vm_Callback_Phone__c']: contact_attributes,
os.environ['sf_Voicemail__c_vm_Connect_Contact_Attributes__c']: international_phone,
os.environ['sf_Voicemail__c_vm_Voicemail_Link__c']: formatted_url

response = sf.create(sobject='Voicemail__c', data=data)

            logger.info('Record {0} Result: Voicemail created [{1}]'.format(loop_counter, response))
            sf_Voicemail__c = response

**Reply by dougjaso** [reply]
> Do you have any thoughts on these errors occurring from the deployment followed by the book to the T? I want to make sure it's working before I introduce any code changes.
> 
> [ERROR] 2022-01-12T20:41:10.284Z b4998231-106c-4c6b-b9fe-a63d86b8baea list indices must be integers or slices, not str [ERROR] 2022-01-12T20:41:10.285Z b4998231-106c-4c6b-b9fe-a63d86b8baea Record 1 Result: Failed to create case
> 
> [ERROR] KeyError: 'Records' Traceback (most recent call last): File "/var/task/awsscv_vmx_packager.py", line 35, in lambda_handler for record in event['Records']:
> 
> Based on my research this is a code-based error or something in previous execution that is not providing the proper input to the packager. Thus, causing a traceback error. Though, I still cannot see any way to resolve the list indices type error.


Encountering this error tells me that you are running a test of the function itself and not passing the correct event in. This function is designed to be triggered by the S3 event when the transcript file is created and dropped into the S3 bucket. When that happens, S3 sends an event to Lambda, containing records of all events that caused the trigger. If you are using the base test event that comes in Lambda by default, you will see this error.

**Reply by dougjaso** [reply]
> Doug,
> 
> You stated that "Whatever field you want to place the attribute list in, that is the value that should be set. Same goes for os.environ['sf_case_vm_phone_field'] (custom phone field in the case object)"
> 
> If I'm understanding correctly, the Voicemail Express Prerequisites doc instructs us to set field labels "Voicemail Link", "Callback Phone", and "Connect Contact Attributes".
> 
> In the case of using the custom object API name Voicemail__c along with the field created per the instructions, wouldn't we want it to be written as such:
> 
> os.environ['sf_Voicemail__c_vm_Callback_Phone__c']: contact_attributes, os.environ['sf_Voicemail__c_vm_Connect_Contact_Attributes__c']: international_phone, os.environ['sf_Voicemail__c_vm_Voicemail_Link__c']: formatted_url
> 
> response = sf.create(sobject='Voicemail__c', data=data)
> 
> ```
>         logger.info('Record {0} Result: Voicemail created [{1}]'.format(loop_counter, response))
>         sf_Voicemail__c = response
> ```

I think you are not understanding what is being referenced here. `os.environ['sf_case_vm_phone_field']` is telling Lambda to reference the hard coded environment variable with the key `sf_case_vm_phone_field` and return that value here. You can see the environment variables in your Lambda function by opening the function then going to Configuration > Environment Variables. These are set during deployment if you named your custom field (at this point in code we do not care what the object itself is) `vm_Callback_Phone__c`, then the proper reference in the code is STILL `os.environ['sf_case_vm_phone_field']`, which returns the field name you provided, `vm_Callback_Phone__c`. This is how the code knows WHERE to insert the phone number in your object.

**Reply by ShinobiSec** [reply]
Now I fully understand. Thank you. 

Do you see the errors in question appropriately being an issue post or more likely something I've configured incorrectly? I conclude that Salesforce and AWS are communicating with each other... correctly... I'll note that the voicemails are being saved and are accessible in the s3 bucket. It appears to me, the record creation in Salesforce is the last leg in the process and is not able to be completed for unknown reasons. I checked the presigner logs since it is the only other thing that functions during/after the recordings and transcripts are saved. All entries state "Found credentials in environment variables.". 

Order of execution of processes was gathered from the Voicemail Express readme.md doc.

**Reply by ShinobiSec** [reply]
I found this document which may track with what we're dealing with. (Outside of general keyerror type situations)

https://stackoverflow.com/questions/41795438/keyerror-records-in-aws-s3-lambda-trigger?rq=1
https://medium.com/simple-thoughts-amplified/passing-variables-from-aws-api-gateway-to-lambda-3c5d8602081b

We do not have access to this part of AWS in our case, therefor I'm still at a loss.

**Comment by dougjaso** [comment, marked as answer]
Providing fix here for anyone else interested. This will be incorporated into a future release.

Two issues here were 1: Writing to an object other than Case and 2: Dealing with the new agent ID structure that uses SF alias instead of the underlying Id of the user. 

To correct for the agent ID portion, the following change was made:
(On a new line after line 175 in vmx_packager):
```
# Validate the agent's Salesforce ID, and correct, if necessary
        if not sf_agent_id.startswith('005'):
            try:
                get_agent_id = sf.query("SELECT Id FROM User WHERE Alias = '" + sf_agent_id + "' LIMIT 1")
                sf_agent_id = get_agent_id[0]['Id']
            except Exception as e:
                logger.error(e)
                logger.error('Record {0} Result: Failed to validate Salesforce User ID'.format(loop_counter))
                continue
```
That checks to see if the already extracted ID begins with 005, which is the default prefix for users. If it does not, we assume an Alias is being used and query based on that Alias. Since Alias is not a unique identifier, we're just grabbing the first one we match and using it, so not duplicating Alias values will be important.

For the second part, we wrote the voicemail to a new custom object named `Voicemail`. Once we had all the right custom fields in the voicemail object, we just changed the code of the write as follows:
```
response = sf.create(sobject='Voicemail__c', data=data)
```
Voicemail delivery was validate for both agents and queue targets.
